Misc fixings and add work thread helper support

lxbsz opened the pull request:
> The following patchset fixes several bugs then adds support to track and report errors to handler specific tools.
> 
> Note:
> 
> I'm not working on tcmu-runner anymore. I'm just posting this with the hope it's useful to the next maintainer. You can give me review comments, but I will not be updating the patchset.

This PR is based Mike's previous https://github.com/open-iscsi/tcmu-runner/pull/626 and with some fixings. 
Thank you @mikechristie  for your work for this.

dillaman commented on main.c:
```diff
@@ -723,6 +725,12 @@ static void check_for_timed_out_cmds(struct tcmu_device *dev)
 		tcmur_cmd->timed_out = true;
+		/*
+		 * These time outs are only currently used for diagnostic
+		 * purpposes right now, so we do not want to escalate the
```
```suggestion
		 * purposes right now, so we do not want to escalate the
```

lxbsz replied:
Sure, will fix it.

dillaman commented on target.c:
```diff
@@ -329,8 +330,13 @@ int tcmu_add_dev_to_recovery_list(struct tcmu_device *dev)
+	pthread_attr_init(&attr);
+	pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);
+
+	ret = pthread_create(&thread_id, &attr,
```
If this tread is just running in the background, how is it tracked for joining before the process exits? Why not re-use the new `tcmur_work` thread wrapper and add the work struct to `tgt_port_grp`?

lxbsz replied:
Yeah, good catch. In linux the detached thread will be terminated when the main process is exiting.  But there still have one problem, if the main process is exiting and it will release all the resources before it terminates all the detached threads, which may still accessing them.

I will convert to the tcmur_work thread wrapper, but the tcmur_work thread wrapper has one bug, I will fix it at the same time.

dillaman commented on tcmur_device.c:
```diff
@@ -489,3 +529,14 @@ void *tcmur_dev_get_private(struct tcmu_device *dev)
+{
+	struct tcmur_device *rdev = tcmu_dev_get_private(dev);
+
+	rdev->cmd_timed_out_cnt++;
```
Nit: is someone holding the `state_lock` lock on `tcmur_device`?

lxbsz commented on tcmur_work.c:
```diff
@@ -0,0 +1,119 @@
+	 * Some handlers will crash if we do a cancel so we just wait.
+	 */
+	tcmu_dbg("waiting for %lu to complete\n", work->thread);
+	pthread_cond_wait(&work->cond, &work->lock);
```
Commit myself:

For the work here we need one refcnt, just think if there has 2 threads queued, and when destroying and flushing the work, the flush could only make sure the first thread is done, the second may still running.

lxbsz commented:
@dillaman 
Updated it, for this version I remove the `work->state` and added one `work->refcnt`, with this we can run and queue more than one threads, the old code can only run and queue one thread.

dillaman commented on tcmur_work.c:
```diff
@@ -0,0 +1,105 @@
+	pthread_mutex_lock(&work->lock);
+	work->work_fn = work_fn;
+	work->data = data;
+	ret = pthread_create(&work->thread, &attr, tcmur_work_fn, work);
```
This would stomp on the `work->thread` of a previously running thread, right?  That makes `__tcmur_flush_work` check of the thread suspicious.

lxbsz replied:
Yeah, right.

I have removed the thread in work and will use the pthread name to identify whether the work thread itself is trying to call the tcmur_flush_work(), which will cause stuck forever.

lxbsz replied:
Together with this versoin, I have remove the lock_work and since we can run multiple thread by using one event_work. And also remove the event type when in event reporting helper, since we are always reporting all the event every time.

dillaman commented on tcmur_device.c:
```diff
@@ -489,3 +542,20 @@ void *tcmur_dev_get_private(struct tcmu_device *dev)
 	return rdev->hm_private;
 }
+
+void tcmu_notify_cmd_timed_out(struct tcmu_device *dev, bool can_access_dev)
```
Nit: is the `can_access_dev` needed? It looks like everyone calls this method w/ `true`

lxbsz replied:
Removed.

lxbsz commented:
Merged, thanks.
